lib/diff.ml

```diff
@@ -123,39 +143,169 @@ let rec type_item ~typing_env ~name ~reference ~current =
       in
       match (type_coercion1 (), type_coercion2 ()) with
       | None, None -> None
-      | _, _ -> (
+      | ref_mismatch, cur_mismatch -> (
           match (reference.type_kind, current.type_kind) with
-          | Type_record (ref_label_lst, _), Type_record (cur_label_lst, _) ->
-              let changed_lbls =
-                modified_record_type ~typing_env ~ref_label_lst ~cur_label_lst
-              in
-              Some
-                (Type { tname = name; tdiff = Modified (Compound changed_lbls) })
-          | _, _ ->
+          | Type_record (ref_label_lst, _), Type_record (cur_label_lst, _) -> (
+              match (ref_mismatch, cur_mismatch) with
+              | ( Some (Includecore.Record_mismatch _),
+                  Some (Includecore.Record_mismatch _) ) ->
+                  let changed_lbls =
+                    modified_record_type ~typing_env ~ref_label_lst
+                      ~cur_label_lst
+                  in
+                  Some
+                    (Type
+                       {
+                         tname = name;
+                         tdiff = Modified (Record_diff changed_lbls);
+                       })
+              | _, _ -> None)
+          | ( Type_variant (ref_constructor_lst, _),
+              Type_variant (cur_constructor_lst, _) ) -> (
+              match (ref_mismatch, cur_mismatch) with
+              | ( Some (Includecore.Variant_mismatch _),
+                  Some (Includecore.Variant_mismatch _) ) ->
+                  let changed_constrs =
+                    modified_variant_type ~typing_env ~ref_constructor_lst
+                      ~cur_constructor_lst
+                  in
+                  Some
+                    (Type
+                       {
+                         tname = name;
+                         tdiff = Modified (Variant_diff changed_constrs);
+                       })
+              | _, _ -> None)
+          | _ ->
               Some
                 (Type
                    {
                      tname = name;
                      tdiff = Modified (Atomic { reference; current });
                    })))
 
-and modified_record_type ~typing_env ~ref_label_lst ~cur_label_lst =
+and modified_variant_type ~typing_env ~ref_constructor_lst ~cur_constructor_lst
+    =
+  let same_args cstr1_args cstr2_args =
+    match (cstr1_args, cstr2_args) with
+    | Cstr_tuple type_lst1, Cstr_tuple type_lst2 -> (
+        try
+          if
+            List.for_all2
```

Comment:
We're going through the entire list twice here: once to tell if there is a diff at all, and a second time to compute the actual diff, calling `does_match` on all elements of the list each time.

I think we can simplify all this by simply computing the diff as a `diff option list` first and simply detecting the case where it's all `None`. That way we call `does_match` only once for each pair.

Reply:
This applies for the record case below as well.